server/src/main/java/org/apache/druid/server/http/MetadataResource.java

```diff
@@ -161,14 +153,11 @@ public Response getDatabaseSegments(
     // If we haven't polled the metadata store yet, use an empty list of datasources.
     Collection<ImmutableDruidDataSource> druidDataSources = Optional.ofNullable(metadataSegmentManager.getDataSources())
                                                                     .orElse(Collections.emptyList());
-
-    if (datasources != null && !datasources.isEmpty()) {
-      druidDataSources = druidDataSources.stream()
-                                         .filter(src -> datasources.contains(src.getName()))
-                                         .collect(Collectors.toSet());
-    }
     final Stream<DataSegment> metadataSegments = druidDataSources
         .stream()
+        .filter((datasources != null && !datasources.isEmpty())
+                ? src -> datasources.contains(src.getName())
+                : src -> true)
```

Comment:
This code is not readable.

Please extract a stream variable and apply a filter only if the condition applies.

Reply:
okay, tried to simplify with your suggestions.